### PR TITLE
Turn `extern crate` and `use` into first-class items.

### PR DESCRIPTION
**Pull request by eddyb**
This implements an important part of RFC PR 385 (see #18219).

### CODE REVIEW

**Comment by eddyb** [comment]
I think I would handle such ordering restrictions in `resolve`, since I expect that's where the proper logic will be -eventually- implemented.
Does that make sense, @alexcrichton?
cc @nikomatsakis

**Comment by Kimundi** [comment]
As far as I know, those ordering restrictions don't apply to items, as those are mutually recursive visible:

```
// In a block:

static FOO ... // visible before and after
use Bar ... // visible before and after
extern crate Baz ... // visible before and after
fn f() ... // visible before and after
let x ... // visible _after_
```

So, what a construct like `let foo; use foo` means in regard to shadowing depends on what items in those positions are supposed to mean in general, and is not specific to view items.

**Comment by eddyb** [comment]
Ah, this causes the weird "only irrefutable patterns allowed here" error (can we stop matching constants in irrefutable patterns already?):
```rust
fn main() {
    let foo = 5u;
    const foo: uint = 42;
    println!("{}", foo);
}
```
And when that is not an issue, it seems that local bindings shadow items:
```rust
fn main() {
    let foo = || 5u;
    fn foo() -> uint { 42 }
    println!("{}", foo()); // prints 5
}
```

**Comment by alexcrichton** [comment]
This is partly why I would like to be conservative here and lift the restriction at a later date. I don't want to allow rearranging items hoping for changes to resolve to land soon, I would rather land the two at the same time.

@Kimundi has a good point though that other items are already plagued by this unfortunate shadowing, so it may not be as dire as I think, but I'd prefer to stem the flow of awkward questions rather than pile more on.

**Comment by Kimundi** [comment]
I guess the forward compatible thing would be to either artificially keep the order restriction in blocks, or to make local bindings with the same name as an item _in the same flat scope_ a hard error.

 ---

This is of course not the place to discuss those things, but I guess the only "sane" semantic for items in blocks in general is the current situation: items are visible everywhere in a scope, but locals, having the property of allowing forward-only shadowing of elements in the same scope, can shadow them.

Otherwise you get weird inconsistent semantic between those two entities:
- items: shadowable with scope only, visible forward and backwards
- locals: shadowable with scope and later locals, visible forward-only

**Comment by eddyb** [comment]
@alexcrichton now that alpha's over, I've rebased this branch and added that error you requested, in the last commit. r?

File: src/libsyntax/feature_gate.rs [hunk 1]
```diff
@@ -257,6 +239,14 @@ impl<'a, 'v> Visitor<'v> for PostExpansionVisitor<'a> {
             }
         }
         match i.node {
+            ast::ItemExternCrate(_) => {
+                if attr::contains_name(&i.attrs[], "phase") {
+                    self.gate_feature("phase", i.span,
+                                      "compile time crate loading is \
+                                       experimental and possibly buggy");
+                }
+            }
```
Comment (alexcrichton) [inline src/libsyntax/feature_gate.rs]: I think this may have snuck back in

File: src/test/auxiliary/blind-item-mixed-crate-use-item-foo.rs [hunk 2]
```diff
@@ -8,8 +8,6 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-pub extern crate core; //~ ERROR: `pub` visibility is not allowed
+#![crate_type="lib"]
 
-fn main() {
-    pub use std::usize; //~ ERROR: imports in functions are never reachable
```
Comment (alexcrichton) [inline src/test/auxiliary/blind-item-mixed-crate-use-item-foo.rs]: Just confirming, but there is still a test for `pub use` in a function being an error, right?

Comment (alexcrichton) [inline src/test/auxiliary/blind-item-mixed-crate-use-item-foo.rs]: Ah I see it below, nevermind!

**Comment by alexcrichton** [comment]
Alright, this all looks great to me, thanks @eddyb! This is a pretty big change though so I would like to bring it up at the next weekly meeting just to make sure everyone's comfortable with pulling the trigger.

File: src/libsyntax/feature_gate.rs (hunk 1 above)
Comment (eddyb) [inline src/libsyntax/feature_gate.rs]: Oops, indeed, a few of these were merged automatically during the rebase.

**Comment by alexcrichton** [comment]
We [chatted about this](https://github.com/rust-lang/meeting-minutes/blob/master/weekly-meetings/2015-01-20.md#goodbye-view-items) in today's meeting, and we all think we're good moving forward. Thanks again for the RFC @Kimundi and the implementation @eddyb!

**Comment by SimonSapin** [comment]
Does this fix https://github.com/rust-lang/rust/issues/18810 ?

**Pull request merged**